test/services/bill-runs/two-part-tariff/allocate-returns-to-charge-element.service.test.js

```diff
@@ -0,0 +1,347 @@
+      })
+
+      it('correctly allocates 10 to the charge element', () => {
+        const { chargeElement, chargeReference, matchingReturns } = testData
+
+        AllocateReturnsToChargeElementService.go(chargeElement, matchingReturns, chargePeriod, chargeReference)
+
+        expect(chargeElement.authorisedAnnualQuantity).to.equal(10)
+        expect(chargeElement.allocatedQuantity).to.equal(10)
+        expect(chargeElement.returnLogs[0].allocatedQuantity).to.equal(10)
+      })
+
+      it('correctly allocates 10 from the return log', () => {
+        const { chargeElement, chargeReference, matchingReturns } = testData
+
+        AllocateReturnsToChargeElementService.go(chargeElement, matchingReturns, chargePeriod, chargeReference)
+
+        expect(matchingReturns[0].allocatedQuantity).to.equal(10)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[0].unallocated).to.equal(0)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[1].unallocated).to.equal(0)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[2].unallocated).to.equal(2)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[3].unallocated).to.equal(4)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[4].unallocated).to.equal(4)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[5].unallocated).to.equal(4)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[6].unallocated).to.equal(4)
+        expect(matchingReturns[0].returnSubmissions[0].returnSubmissionLines[7].unallocated).to.equal(4)
+      })
+    })
+
+    describe('with a `returnSubmissionLine` outside of the `chargePeriod`', () => {
+      beforeEach(() => {
+        testData = _generateTestData()
+        testData.matchingReturns[0].returnSubmissions[0].returnSubmissionLines[0].startDate = new Date('2022-03-01')
+        testData.matchingReturns[0].returnSubmissions[0].returnSubmissionLines[0].endDate = new Date('2022-03-30')
+      })
+
+      it('correctly allocates 28 to the charge reference', () => {
+        const { chargeElement, chargeReference, matchingReturns } = testData
+
+        AllocateReturnsToChargeElementService.go(chargeElement, matchingReturns, chargePeriod, chargeReference)
+
+        expect(chargeReference.allocatedQuantity).to.equal(28)
+      })
+
+      it('correctly allocates 28 to the charge element', () => {
+        const { chargeElement, chargeReference, matchingReturns } = testData
+
+        AllocateReturnsToChargeElementService.go(chargeElement, matchingReturns, chargePeriod, chargeReference)
+
+        expect(chargeElement.allocatedQuantity).to.equal(28)
+        expect(chargeElement.returnLogs[0].allocatedQuantity).to.equal(28)
+      })
+
+      it('does not include the line outside of the charge period and allocates 28 from the return log', () => {
```

Comment:
Do we need to add here that the chargeDatesOverlap would be true?

Reply:
nope 😀